cpp/src/io/parquet/writer_impl.cu

```diff
@@ -0,0 +1,691 @@
+
+#include <cudf/null_mask.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+
+#include <algorithm>
+#include <cstring>
+#include <utility>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/device_buffer.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace io {
+namespace detail {
+namespace parquet {
+
+using namespace cudf::io::parquet;
+using namespace cudf::io;
+
+namespace {
+
+/**
+ * @brief Helper for pinned host memory
+ **/
+template <typename T>
+using pinned_buffer = std::unique_ptr<T, decltype(&cudaFreeHost)>;
+
+/**
+ * @brief Function that translates GDF compression to parquet compression
+ **/
+parquet::Compression to_parquet_compression(
+    compression_type compression) {
+  switch (compression) {
+    case compression_type::AUTO:
+    case compression_type::SNAPPY:
+      return parquet::Compression::SNAPPY;
+    case compression_type::NONE:
+      return parquet::Compression::UNCOMPRESSED;
+    default:
+      CUDF_EXPECTS(false, "Unsupported compression type");
+      return parquet::Compression::UNCOMPRESSED;
+  }
+}
+
+}  // namespace
+
+
+/**
+ * @brief Helper kernel for converting string data/offsets into nvstrdesc
+ * REMOVEME: Once we eliminate the legacy readers/writers, the kernels could be
+ * made to use the native offset+data layout.
+ **/
+__global__ void stringdata_to_nvstrdesc(gpu::nvstrdesc_s *dst, const size_type *offsets,
+                        const char *strdata, const uint32_t *nulls,
+                        size_type column_size) {
+  size_type row = blockIdx.x * blockDim.x + threadIdx.x;
+  if (row < column_size) {
+    uint32_t is_valid = (nulls) ? (nulls[row >> 5] >> (row & 0x1f)) & 1 : 1;
+    size_t count;
+    const char *ptr;
+    if (is_valid) {
+      size_type cur = offsets[row];
+      size_type next = offsets[row + 1];
+      ptr = strdata + cur;
+      count = (next > cur) ? next - cur : 0;
+    }
+    else {
+      ptr = nullptr;
+      count = 0;
+    }
+    dst[row].ptr = ptr;
+    dst[row].count = count;
+  }
+}
+
+
+/**
+ * @brief Helper class that adds parquet-specific column info
+ **/
+class parquet_column_view {
+ public:
+  /**
+   * @brief Constructor that extracts out the string position + length pairs
+   * for building dictionaries for string columns
+   **/
+  explicit parquet_column_view(size_t id, column_view const &col, cudaStream_t stream)
+      : _id(id),
+        _string_type(col.type().id() == type_id::STRING),
+        _type_width(_string_type ? 0 : cudf::size_of(col.type())),
+        _converted_type(ConvertedType::UNKNOWN),
+        _ts_scale(0),
+        _data_count(col.size()),
+        _null_count(col.null_count()),
+        _data(col.data<uint8_t>()),
+        _nulls(col.has_nulls() ? col.null_mask() : nullptr) {
+    switch(col.type().id()) {
+      case cudf::type_id::INT8:
+        _physical_type = Type::INT32;
+        _converted_type = ConvertedType::INT_8;
+        _stats_dtype = statistics_dtype::dtype_int8;
```

Comment:
Now that the statistics are grabbing 4 bytes for int8 and int16, do we need to make sure this type is also that width?  Otherwise wondering if we're currently relying on bytes beyond these types to be zero and if they may not be guaranteed to be zero.

Reply:
We're good: the data in the statistics header is written from the statistics_val type buffer, which is always int64 for int types (regardless of the column data type in memory). For example, with int8 types, the int64 value is a sign-extended version of the 8-bit value.